pyspectrafuse/common/noid_cluster_db.py

```
@@ -0,0 +1,297 @@
+def _membership_stats(membership_path: str) -> pd.DataFrame:
+    conn = duckdb.connect(":memory:")
+    try:
+        return conn.execute(f"""
+            SELECT
+                cluster_id,
+                CAST(COUNT(*) AS INTEGER) AS member_count,
+                CAST(COUNT(DISTINCT project_accession) AS SMALLINT) AS project_count,
+                list_distinct(array_agg(project_accession)) AS source_datasets
+            FROM read_parquet('{membership_path}')
+            GROUP BY cluster_id
+        """).fetchdf()
```

Ruff (0.15.12)

[error] Possible SQL injection vector through string-based query construction

(S608)